File: tests/bridges/lua_scripts/t_echo.lua
local atp = require("atp")

local M = atp.module("lua_echo", { 1, 0 })

M.in_i32 = atp.input(atp.i32)
M.in_i64 = atp.input(atp.i64)
M.in_f64 = atp.input(atp.f64)
M.in_bool = atp.input(atp.bool)
M.in_text = atp.input(atp.text)
M.in_blob = atp.input(atp.blob)

M.out_i32 = atp.output(atp.i32)
M.out_i64 = atp.output(atp.i64)
M.out_f64 = atp.output(atp.f64)
M.out_bool = atp.output(atp.bool)
M.out_text = atp.output(atp.text)
M.out_blob = atp.output(atp.blob)
M.out_gain = atp.output(atp.f64)

M.gain = atp.property(atp.f64, 1.0)
M.overflow = atp.property(atp.bool, false)

function M:iterate()
    local n = self.in_i32:take()
    if n == nil then
        return atp.IDLE
    end
    if self.overflow:get() then
        self.out_i32:write(2147483648)
        return atp.BUSY
    end
    self.out_i32:write(n)
    self.out_gain:write(n * self.gain:get())

    local wide = self.in_i64:take()
    if wide ~= nil then
        self.out_i64:write(wide)
    end
    local number = self.in_f64:take()
    if number ~= nil then
        self.out_f64:write(number)
    end
    local flag = self.in_bool:take()
    if flag ~= nil then
        self.out_bool:write(not flag)
    end
    local text = self.in_text:take()
    if text ~= nil then
        self.out_text:write(text .. "!")
    end
    local bytes = self.in_blob:take()
    if bytes ~= nil then
        self.out_blob:write(string.reverse(bytes))
    end
    return atp.BUSY
end

return M

File: tests/bridges/scripts/t_echo.py
"""Echoes every payload kind back out, so one pass exercises the whole conversion table."""

import atp


class Echo(atp.Module):
    name = "py_echo"

    in_i32 = atp.Input(atp.i32)
    in_i64 = atp.Input(atp.i64)
    in_f64 = atp.Input(atp.f64)
    in_bool = atp.Input(atp.bool_)
    in_text = atp.Input(atp.text)
    in_blob = atp.Input(atp.blob)

    out_i32 = atp.Output(atp.i32)
    out_i64 = atp.Output(atp.i64)
    out_f64 = atp.Output(atp.f64)
    out_bool = atp.Output(atp.bool_)
    out_text = atp.Output(atp.text)
    out_blob = atp.Output(atp.blob)
    out_gain = atp.Output(atp.f64)

    gain = atp.Property(atp.f64, 1.5)
    overflow = atp.Property(atp.bool_, False, persistent=False)

    def initialize(self):
        self.passes = 0
        self.log("py_echo ready")

    def _echo(self, source, sink, transform=None):
        value = source.get()
        if value is None:
            return
        sink.write(value if transform is None else transform(value))

    def iterate(self):
        value = self.in_i32.get()
        if value is None:
            return atp.IDLE
        self.passes += 1
        if self.overflow.get():
            self.out_i32.write(2**31)
            return atp.BUSY
        self.out_i32.write(value)
        self._echo(self.in_i64, self.out_i64)
        self._echo(self.in_f64, self.out_f64)
        self._echo(self.in_bool, self.out_bool)
        self._echo(self.in_text, self.out_text, lambda seen: seen + "!")
        self._echo(self.in_blob, self.out_blob, lambda seen: bytes(reversed(seen)))
        self.out_gain.write(self.gain.get())
        return atp.BUSY

    def stop(self):
        self.passes = 0
